File: sample/src/main/java/ru/terrakok/cicerone/sample/mvp/animation/profile/ProfilePresenter.java
package ru.terrakok.cicerone.sample.mvp.animation.profile;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;

import ru.terrakok.cicerone.Router;
import ru.terrakok.cicerone.sample.Screens;
import ru.terrakok.cicerone.sample.mvp.animation.PhotoSelection;

/**
 * Created by Konstantin Tskhovrebov (aka @terrakok) on 14.07.17.
 */

@InjectViewState
public class ProfilePresenter extends MvpPresenter<ProfileView> {

    private Router router;
    private PhotoSelection photoSelection;

    public ProfilePresenter(PhotoSelection photoSelection, Router router) {
        this.photoSelection = photoSelection;
        this.router = router;

        photoSelection.setListener(new PhotoSelection.Listener() {
            @Override
            public void onChange(int selectedPhoto) {
                updatePhoto();
            }
        });
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        updatePhoto();
    }

    @Override
    public void onDestroy() {
        photoSelection.setListener(null);
        super.onDestroy();
    }

    private void updatePhoto() {
        getViewState().showPhoto(photoSelection.getSelectedPhoto());
    }

    public void onPhotoClicked() {
        router.navigateTo(new Screens.SelectPhotoScreen());
    }

    public void onBackPressed() {
        router.exit();
    }
}

File: sample/src/main/java/com/github/terrakok/cicerone/sample/mvp/animation/profile/ProfilePresenter.kt
package com.github.terrakok.cicerone.sample.mvp.animation.profile

import com.github.terrakok.cicerone.ResultListenerHandler
import com.github.terrakok.cicerone.Router
import com.github.terrakok.cicerone.sample.R
import com.github.terrakok.cicerone.sample.Screens.SelectPhoto
import moxy.InjectViewState
import moxy.MvpPresenter

/**
 * Created by Konstantin Tskhovrebov (aka @terrakok) on 14.07.17.
 */
@InjectViewState
class ProfilePresenter(
        private val router: Router
) : MvpPresenter<ProfileView>() {
    private var resultListenerHandler: ResultListenerHandler? = null

    companion object {
        private const val RESULT_KEY = "photo_result"
        private const val DEFAULT_PHOTO = R.drawable.ava_1
    }

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState!!.showPhoto(DEFAULT_PHOTO)
    }

    fun onPhotoClicked() {
        resultListenerHandler = router.setResultListener(RESULT_KEY) { data ->
            viewState!!.showPhoto(data as Int)
        }
        router.navigateTo(SelectPhoto(RESULT_KEY))
    }

    override fun onDestroy() {
        resultListenerHandler?.dispose()
        super.onDestroy()
    }

    fun onBackPressed() {
        router.exit()
    }
}